sympy/polys/matrices/normalforms.py

```diff
@@ -59,98 +59,194 @@ def invariant_factors(m):
+    smf = DomainMatrix.diag(invs, domain, shape).to_dense()
+
+    s = DomainMatrix(s, domain=domain, shape=(rows, rows))
+    t = DomainMatrix(t, domain=domain, shape=(cols, cols))
+    return smf, s, t
+
+
+def _smith_normal_decomp(m, domain, shape, full):
+    '''
+    Return the tuple of abelian invariants for a matrix `m`
+    (as in the Smith-Normal form). If `full=True` then invertible matrices
+    ``s, t`` such that the product ``s, m, t`` is the Smith Normal Form
+    are also returned.
+    '''
     if not domain.is_PID:
         msg = "The matrix entries must be over a principal ideal domain"
         raise ValueError(msg)
 
-    if 0 in m.shape:
-        return ()
+    rows, cols = shape
 
-    rows, cols = shape = m.shape
-    m = list(m.to_dense().rep.to_ddm())
+    def eye(n):
+        return [[domain.one if i == j else domain.zero for i in range(n)] for j in range(n)]
+
+    if 0 in shape:
+        if full:
+            return (), eye(rows), eye(cols)
+        else:
+            return ()
+
+    if full:
+        s = eye(rows)
+        t = eye(cols)
 
     def add_rows(m, i, j, a, b, c, d):
         # replace m[i, :] by a*m[i, :] + b*m[j, :]
         # and m[j, :] by c*m[i, :] + d*m[j, :]
-        for k in range(cols):
+        for k in range(len(m[0])):
             e = m[i][k]
             m[i][k] = a*e + b*m[j][k]
             m[j][k] = c*e + d*m[j][k]
 
-    def clear_column(m):
+    def clear_column():
         # make m[1:, 0] zero by row and column operations
         if m[0][0] == 0:
-            return m  # pragma: nocover
+            return  # pragma: nocover
         pivot = m[0][0]
         for j in range(1, rows):
             if m[j][0] == 0:
                 continue
             d, r = domain.div(m[j][0], pivot)
             if r == 0:
                 add_rows(m, 0, j, 1, 0, -d, 1)
+                if full:
+                    add_rows(s, 0, j, 1, 0, -d, 1)
             else:
                 a, b, g = domain.gcdex(pivot, m[j][0])
                 d_0 = domain.div(m[j][0], g)[0]
                 d_j = domain.div(pivot, g)[0]
                 add_rows(m, 0, j, a, b, d_0, -d_j)
+                if full:
+                    add_rows(s, 0, j, a, b, d_0, -d_j)
                 pivot = g
-        return m
 
-    def clear_row(m):
+    def clear_row():
         # make m[0, 1:] zero by row and column operations
-        if m[0][0] == 0:
-            return m  # pragma: nocover
         pivot = m[0][0]
         for j in range(1, cols):
             if m[0][j] == 0:
                 continue
             d, r = domain.div(m[0][j], pivot)
             if r == 0:
                 add_columns(m, 0, j, 1, 0, -d, 1)
+                if full:
+                    add_columns(t, 0, j, 1, 0, -d, 1)
             else:
                 a, b, g = domain.gcdex(pivot, m[0][j])
                 d_0 = domain.div(m[0][j], g)[0]
                 d_j = domain.div(pivot, g)[0]
```

Comment:
Is this intended to be exact division or is there a nonzero remainder that is being discarded?

If so it is typically better to use `domain.exquo`.

Reply:
Using `exquo` now